File: src/main/java/com/bajdcc/LALR1/syntax/exp/OptionExp.java
package com.bajdcc.LALR1.syntax.exp;

import com.bajdcc.LALR1.syntax.ISyntaxComponent;
import com.bajdcc.LALR1.syntax.ISyntaxComponentVisitor;
import com.bajdcc.util.VisitBag;

/**
 * 文法规则（可选）
 *
 * @author bajdcc
 */
public class OptionExp implements ISyntaxComponent {

	/**
	 * 子表达式
	 */
	public ISyntaxComponent expression = null;

	@Override
	public void visit(ISyntaxComponentVisitor visitor) {
		VisitBag bag = new VisitBag();
		visitor.visitBegin(this, bag);
		if (bag.getVisitChildren()) {
			expression.visit(visitor);
		}
		if (bag.getVisitEnd()) {
			visitor.visitEnd(this);
		}
	}
}

File: src/main/java/com/bajdcc/LALR1/syntax/exp/OptionExp.kt
package com.bajdcc.LALR1.syntax.exp

import com.bajdcc.LALR1.syntax.ISyntaxComponent
import com.bajdcc.LALR1.syntax.ISyntaxComponentVisitor
import com.bajdcc.util.VisitBag

/**
 * 文法规则（可选）
 * [expression] 子表达式
 * @author bajdcc
 */
class OptionExp(var expression: ISyntaxComponent) : ISyntaxComponent {

    override fun visit(visitor: ISyntaxComponentVisitor) {
        val bag = VisitBag()
        visitor.visitBegin(this, bag)
        if (bag.visitChildren) {
            expression.visit(visitor)
        }
        if (bag.visitEnd) {
            visitor.visitEnd(this)
        }
    }
}
